### PR TITLE
Fix/check extension running scripts

### PR DESCRIPTION
**Pull request by misslivirose**
This is a fix for FogBugz case 4816: No filtering of urls in script pane

Previously, ScriptEngine::loadURL didn't filter or check the file extension to ensure that non-script files were not attempted to be loaded when loaded from a URL. This fix checks that the script is valid JSON or JS before loading and if it isn't a supported file type, alerts the user to the error.

Manual test plan: Attempt to run script files from URL, adding in a few non-script file types. Ensure that the proper extension scripts run, while non-JS/JSON files do not.

Scripted test plan: run the following script and look for the correct output:  https://hifi-content.s3.amazonaws.com/liv/dev/TestScripts/testScriptLoad.js 

Test 1: Hello.js, prints to "Hello, Test 1" script debug window
Test 2: Hello.json, prints "Hello, Test 2" to script debug window
Test 3: Tests capitalized file extensions, prints "Hello, Test 3" to script debug window
Test 4: Tests url with additional '.' breaks in name, prints "Hello, Test 4" to script debug window
Test 5: Tests url with appended Date.now(), prints "Hello, Test 4" a second time to script debug window
Test 6: Incorrect file type, error UI shown w/ error message in log ([CRITICAL] [hifi.scriptengine] "File extension of file: https://hifi-content.s3.amazonaws.com/liv/dev/Silver_star.png is not a currently supported script type")
Test 7: Confirm that vec 3 (no file extension) loads properly, prints a vector to script debug window

This should not impact global modules that do not have a file extension, as those do not pass through loadURL (which is included in the test script by way of the vec3 module) but further confirmation of that when they are more widely used could be beneficial.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/131/HighFidelity-Beta-PR10536-09c7a615993a2701b32157887e70bffd4bf3aa25.exe

File: libraries/script-engine/src/ScriptEngine.cpp [hunk 1]
```diff
@@ -425,6 +425,12 @@ QString ScriptEngine::getFilename() const {
     return lastPart;
 }
 
+bool ScriptEngine::hasValidScriptSuffix(const QString& scriptFileName) {
+	QFileInfo fileInfo(scriptFileName);
```
Comment (ZappoMan) [inline libraries/script-engine/src/ScriptEngine.cpp]: coding standard: indent is wrong, should be 4 spaces per indent.

File: libraries/script-engine/src/ScriptEngine.h [hunk 2]
```diff
@@ -146,6 +146,7 @@ class ScriptEngine : public BaseScriptEngine, public EntitiesScriptEngineProvide
     /// to run... NOTE - this is used by Application currently to load the url. We don't really want it to be exposed
     /// to scripts. we may not need this to be invokable
     void loadURL(const QUrl& scriptURL, bool reload);
+	bool hasValidScriptSuffix(const QString& scriptFileName);
```
Comment (ZappoMan) [inline libraries/script-engine/src/ScriptEngine.h]: coding standard: indent wrong

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/131/HighFidelity-Beta-PR10536-09c7a615993a2701b32157887e70bffd4bf3aa25.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/132/HighFidelity-Beta-PR10536-b2bbf21fb004fac6652351a9b7d7bb111fa0dd1f.exe

**Comment by CainFoool** [comment]
Will test.

**Comment by CainFoool** [comment]
# Test Failed 

- Test 1: Hello.js, prints to "Hello, Test 1" script debug window - Fail.
- Test 2: Hello.json, prints "Hello, Test 2" to script debug window - Fail.
- Test 3: Tests capitalized file extensions, prints "Hello, Test 3" to script debug window - Pass.
- Test 4: Tests url with additional '.' breaks in name, prints "Hello, Test 4" to script debug window - Pass.
- Test 5: Tests url with appended Date.now(), prints "Hello, Test 4" a second time to script debug window
- Test 6: Incorrect file type, error UI shown w/ error message in log ([CRITICAL] [hifi.scriptengine] "File extension of file: https://hifi-content.s3.amazonaws.com/liv/dev/Silver_star.png is not a currently supported script type") - Pass.
- Test 7: Confirm that vec 3 (no file extension) loads properly, prints a vector to script debug window - Fail.

Test Results (via Log File):
`Done downloading script at: "https://hifi-content.s3.amazonaws.com/liv/dev/TestScripts/testScriptLoad.js"`
`[05/23 23:54:53] [DEBUG] [hifi.scriptengine] loadURL "https://hifi-content.s3.amazonaws.com/liv/dev/TestScripts/testScriptLoad.js" "Success" QThread(0x20a93fb3700, name = "Main Thread")`
`[05/23 23:54:53] [INFO] [hifi.scriptengine] "Script Engine starting:testScriptLoad.js"`
`[05/23 23:54:53] [CRITICAL] [hifi.scriptengine] File extension of file: https://hifi-content.s3.amazonaws.com/liv/dev/Silver_star.png is not a currently supported script type`
`[05/23 23:54:53] [DEBUG] [hifi.scriptengine] Application::loadScript(), script failed to load...`
`[05/23 23:54:53] [DEBUG] [hifi.scriptengine.module] ScriptEngine::require( "vec3" )`
`[05/23 23:54:53] [DEBUG] [hifi.scriptengine.module] newModule "file:///C:/Program Files/High Fidelity - PR10536/scripts/modules/vec3.js" ""`
`[05/23 23:54:53] [DEBUG] [hifi.scriptengine.module] require.fetchModuleSource:  "vec3.js" QThread(0x20acf12f1f0, name = "js:testScriptLoad.js")`
`[05/23 23:54:53] [DEBUG] [hifi.scriptengine] Loading script at  QUrl("file:///C:/Program Files/High Fidelity - PR10536/scripts/modules/vec3.js")`
`[05/23 23:54:53] [DEBUG] [hifi.scriptengine] Done downloading script at: "file:///C:/Program Files/High Fidelity - PR10536/scripts/modules/vec3.js"`
`[05/23 23:54:53] [DEBUG] [hifi.scriptengine] Loaded:  "file:///C:/Program Files/High Fidelity - PR10536/scripts/modules/vec3.js"`
`[05/23 23:54:53] [DEBUG] [hifi.scriptengine.module] require.fetchModuleSource.onload:  "vec3.js" "Success" QThread(0x20acf12f1f0, name = "js:testScriptLoad.js")`
`[05/23 23:54:53] [DEBUG] [hifi.scriptengine.module] "require.instantiateModule: vec3.js / 2243 bytes"`
`[05/23 23:54:53] [DEBUG] [hifi.scriptengine.module] //ScriptEngine::require( "vec3" )`
`[05/23 23:54:53] [DEBUG] [hifi.scriptengine.script] [Vec3 (100.000,200.000,300.000)]`
`[05/23 23:54:54] [DEBUG] [hifi.physics] ObjectConstraintHinge::getConstraint -- no rigidBodyA`
`[05/23 23:54:54] [DEBUG] [hifi.scriptengine] Done downloading script at: "https://hifi-content.s3.amazonaws.com/liv/dev/TestScripts/printHello.js"`
`[05/23 23:54:54] [DEBUG] [hifi.scriptengine] loadURL "https://hifi-content.s3.amazonaws.com/liv/dev/TestScripts/printHello.js" "Success" QThread(0x20a93fb3700, name = "Main Thread")`

`[05/23 23:54:54] [DEBUG] [hifi.scriptengine] Done downloading script at: "https://hifi-content.s3.amazonaws.com/liv/dev/TestScripts/printHello.json"`
`[05/23 23:54:54] [DEBUG] [hifi.scriptengine] loadURL "https://hifi-content.s3.amazonaws.com/liv/dev/TestScripts/printHello.json" "Success" QThread(0x20a93fb3700, name = "Main Thread")`
`[05/23 23:54:54] [DEBUG] [hifi.scriptengine] Done downloading script at: "https://hifi-content.s3.amazonaws.com/liv/dev/TestScripts/printHelloCaps.JS"`
`[05/23 23:54:54] [DEBUG] [hifi.scriptengine] loadURL "https://hifi-content.s3.amazonaws.com/liv/dev/TestScripts/printHelloCaps.JS" "Success" QThread(0x20a93fb3700, name = "Main Thread")`
`[05/23 23:54:54] [INFO] [hifi.scriptengine] "Script Engine starting:printHelloCaps.JS"`
`[05/23 23:54:54] [DEBUG] [hifi.scriptengine.script] Hello, Test 3`
`[05/23 23:54:54] [DEBUG] [hifi.interface.deadlock] DEADLOCK WATCHDOG WARNING: lastHeartbeatAge: 6535 elapsedMovingAverage: 1072 maxElapsed: 4022415 PREVIOUS maxElapsedAverage: 1071 NEW maxElapsedAverage: 1072 ** NEW MAX ELAPSED AVERAGE ** samples: 5793`
`[05/23 23:54:54] [DEBUG] [hifi.scriptengine] Done downloading script at: "https://hifi-content.s3.amazonaws.com/liv/dev/TestScripts/printHello %285.23.2017 2.32.47 PM%29.js?1495580093870"`
`[05/23 23:54:54] [DEBUG] [hifi.scriptengine] loadURL "https://hifi-content.s3.amazonaws.com/liv/dev/TestScripts/printHello %285.23.2017 2.32.47 PM%29.js?1495580093870" "Success" QThread(0x20a93fb3700, name = "Main Thread")`
`[05/23 23:54:54] [INFO] [hifi.scriptengine] "Script Engine starting:printHello %285.23.2017 2.32.47 PM%29.js?1495580093870"`
`[05/23 23:54:54] [DEBUG] [hifi.scriptengine.script] Hello, Test 4`
`[05/23 23:54:54] [DEBUG] [hifi.scriptengine] Done downloading script at: "https://hifi-content.s3.amazonaws.com/liv/dev/TestScripts/printHello %285.23.2017 2.32.47 PM%29.js"`
`[05/23 23:54:54] [DEBUG] [hifi.scriptengine] loadURL "https://hifi-content.s3.amazonaws.com/liv/dev/TestScripts/printHello %285.23.2017 2.32.47 PM%29.js" "Success" QThread(0x20a93fb3700, name = "Main Thread")`
`[05/23 23:54:54] [INFO] [hifi.scriptengine] "Script Engine starting:printHello %285.23.2017 2.32.47 PM%29.js"`
`[05/23 23:54:54] [DEBUG] [hifi.scriptengine.script] Hello, Test 4``

**Comment by misslivirose** [comment]
Can you confirm that you have the script debug window open when running this in addition to the log, and that nothing was showing up for those tests? I'm not sure why the print statements are coming through in the log for some of those, but the scripts seem to be downloading properly so that may be an unrelated bug to this PR.

**Comment by CainFoool** [comment]
@misslivirose In the Script Log this is what popped up:

`[4/24 0:42:32] [testScriptLoad.js] INFO - Script Engine starting:testScriptLoad.js`
`[4/24 0:42:33] [testScriptLoad.js] [Vec3 (100.000,200.000,300.000)]`
`[4/24 0:42:33] [printHello.js] INFO - Script Engine starting:printHello.js`
`[4/24 0:42:33] [printHello.js] Hello, Test 1`
`[4/24 0:42:33] [printHello.json] INFO - Script Engine starting:printHello.json`
`[4/24 0:42:33] [printHello.json] Hello, Test 2`
`[4/24 0:42:33] [printHelloCaps.JS] INFO - Script Engine starting:printHelloCaps.JS`
`[4/24 0:42:33] [printHelloCaps.JS] Hello, Test 3`
`[4/24 0:42:33] [printHello %285.23.2017 2.32.47 PM%29.js] INFO - Script Engine starting:printHello %285.23.2017 2.32.47 PM%29.js`
`[4/24 0:42:33] [printHello %285.23.2017 2.32.47 PM%29.js] Hello, Test 4`
`[4/24 0:42:33] [printHello %285.23.2017 2.32.47 PM%29.js?1495582952998] INFO - Script Engine starting:printHello %285.23.2017 2.32.47 PM%29.js?1495582952998`
`[4/24 0:42:33] [printHello %285.23.2017 2.32.47 PM%29.js?1495582952998] Hello, Test 4`

**Comment by misslivirose** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/136/HighFidelity-Beta-PR10536-b2bbf21fb004fac6652351a9b7d7bb111fa0dd1f.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/136/HighFidelity-Beta-PR10536-b2bbf21fb004fac6652351a9b7d7bb111fa0dd1f.dmg

**Pull request merged**